apps/shipments/admin.py

```diff
@@ -123,4 +135,208 @@ def has_change_permission(self, request, obj=None):
+            self.get_readonly_fields(request, obj),
+            model_admin=self,
+        )
+
+        model_name = original_opts.model_name
+        url_triplet = self.admin_site.name, original_opts.app_label, model_name
+        context = {
+            "title": _("Revert %s") % force_text(obj),
+            "adminform": admin_form,
+            "object_id": object_id,
+            "original": obj,
+            "is_popup": False,
+            "media": mark_safe(self.media + admin_form.media),
+            "errors": helpers.AdminErrorList(form, formsets),
+            "app_label": original_opts.app_label,
+            "original_opts": original_opts,
+            "changelist_url": reverse("%s:%s_%s_changelist" % url_triplet),
+            "change_url": reverse("%s:%s_%s_change" % url_triplet, args=(obj.pk,)),
+            "history_url": reverse("%s:%s_%s_history" % url_triplet, args=(obj.pk,)),
+            "change_history": change_history,
+            # Context variables copied from render_change_form
+            "add": False,
+            "change": True,
+            "has_add_permission": self.has_add_permission(request),
+            "has_change_permission": self.has_change_permission(request, obj),
+            "has_delete_permission": self.has_delete_permission(request, obj),
+            "has_file_field": True,
+            "has_absolute_url": False,
+            "form_url": "",
+            "opts": model._meta,
+            "content_type_id": ContentType.objects.get_for_model(self.model).id,
+            "save_as": self.save_as,
+            "save_on_top": self.save_on_top,
+            "root_path": getattr(self.admin_site, "root_path", None),
+        }
+        context.update(self.admin_site.each_context(request))
+        context.update(extra_context or {})
+        extra_kwargs = {}
+        return render(
+            request, self.object_history_form_template, context, **extra_kwargs
+        )
+
+
+class HistoricalShipmentAdmin(BaseModelHistory, ShipmentAdmin):
+    pass
+
+
+class ShallowUserAdmin(admin.ModelAdmin):
+
+    def has_delete_permission(self, request, obj=None):
+        return False
+
+    def has_add_permission(self, request):
+        return False
+
+    def has_change_permission(self, request, obj=None):
+        return False
+
+
+EXCLUDED_FIELDS = ('shipment_to', 'shipment_from', 'shipment_dest', 'shipment_bill', 'shipment_bill', 'updated_at',
+                   'created_at')
+
+
+class LocationAdmin(BaseModelHistory):
+    fieldsets = [(None, {'fields': [field.name for field in Location._meta.get_fields() if
+                                    field.name not in EXCLUDED_FIELDS]})]
+
+    def has_delete_permission(self, request, obj=None):
+        return False
+
+    def has_add_permission(self, request):
+        return False
+
+    def has_change_permission(self, request, obj=None):
+        return False
+
+
 admin.site.register(Shipment, ShipmentAdmin)
+
+admin.site.unregister(Shipment)
```

Comment:
Why do you register an Admin model, unregister it, and register it with another handler?